[WIP] Column search

mklein994 opened the pull request:
Add search inputs for each column header.

### Things to do before merging
- [ ] Implement in [`GtCustomComponentFactory`](https://github.com/hjalmers/angular-generic-table/blob/fd98f6c107200d8840351c7c956ac1f1b8907e87/%40angular-generic-table/core/components/gt-custom-component-factory.ts#L16-L17)
- [x] Move column search demo module to ~`src/app/demos/`~ `src/app/examples/search/`
- [ ] Make styling more consistent with the rest of the app
- [x] Filter rows by each column's search box (if enabled)
- [ ] Highlight search matches
- [ ] Handle `(keyup)` event asynchronously (maybe?)
- [x] Ability to show only null values mklein994/angular-generic-table#1
- [ ] Search by types other than string mklein994/angular-generic-table#2
- [ ] Split search terms like how it's done in global search
    - For example: searching for `smith brian` under a 'name' field, where first and last names are joined with a space, should find "Brian Smith" if it exists. If the query is wrapped in double quotes, it's counted as one word, so a search `"smith brian"` would not find "Brian Smith".
    - **TODO**: what about case sensitivity?

Closes #99

mklein994 committed:
Add a column search component

mklein994 committed:
Add a mock generic table example for column search

mklein994 commented:
I've labeled commits that shouldn't be merged into master with `DO NOT MERGE:`, to make them easier to find later. These include things like changing travis' build scripts to be specific to this PR, and other minor things. When this PR is ready, I'll rebase those out.

mklein994 commented:
For some reason, I can't seem to get it to work on angular 4, but if I upgrade to angular 5 it works. `ng-packgr` is failing silently when running `npm run build:core`.

I'm going to clean this branch up so that others can contribute to it more easily. (That way I can rebase only when master get's updated).

hjalmers commented:
Hi @mklein994 nice work, I'm also in the middle of refactoring the examples and how they are generated as that's one of the main things that have prevented me from bumping the code to angular 5, once I'm done with that it should be easier for others to contribute I hope. I haven't done anything changes to the lib so it should be pretty easy to merge, feel free to check it out if you like: https://github.com/hjalmers/angular-generic-table/tree/feat/angular-5-update

I've just added one example using the new "format" which is available under '/demo/basic'.

mklein994 commented:
@hjalmers I left a comment over [here](https://github.com/hjalmers/angular-generic-table/commit/53ca9340d5bd51e747da9aa164910b6bdf02f031#commitcomment-28640240) about some of those updates.

mklein994 committed:
Add the gt prefix to column-search selector

mklein994 committed:
Add a column search pipe

mklein994 committed:
Add a searchBox property in gtSettings

This enables a search box for that header.

A search field in the generic-table component was also added.

mklein994 committed:
Add a column search interface to handle search values

This will hold the object keys as properties, and strings as their
values. This may not be the best way to implement it, because the
interface just has an [index signature][1] for the column names, but it
seems to work for now. Hopefully there just get referenced by objectKey.

[1]: https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks

mklein994 committed:
Start implementing column search

WIP; doesn't actually filter rows yet, and it kinda spams the debug log.

mklein994 commented:
I rebased onto master, updating each commit to match the new style.

mklein994 commented:
Rebase onto master (cypress tests).

mklein994 committed:
Replace column search object when updated

This allows angular to detect a change in the pipe.

mklein994 committed:
Implement column search

Replaces the column search array every time a column search box is
updated.

Works so far as long as the field values can be cast as strings.

Also adds the column search pipe to the other chains that include
gtSearch and gtFilter pipes.

mklein994 commented:
@hjalmers OK: so this works... when the table is small. I took your advice wrote it using pipes, because that seemed to be the simplest option to extend it. Everything is performed synchronously, so when I enabled `searchBox` for some fields in the advanced demo, performance was taking a big hit, as you predicted. (It was also failing because I haven't implemented the `search` or `value` transformations yet).

If I were to re-implement this using observables and triggering it asynchronously, what would need to be updated?

hjalmers commented:
Yeah, I first wrote the search function for AngularJs which had a lot of issues with performance due to two-way binding so I tried to minimize the amount of operations performed and just trigger one change and it becomes even trickier when involving multiple columns. I think a lot of the code from the global search could be reused however, but instead of merging all the columns into one string, it could go through the data one column at a time and only go through columns that are included in the search "object". The next iteration should only have to go through the rows that pass the first column check and so on, it shouldn't have to have have that big impact on performance.

It would be a lot nicer to use observables however but I think it would be hard to use observables just for this with out rewriting the other pipes too, I think it would be better to start from scratch and think about the flow which should be something like this:

1. filter
2. global search
3. column search (might be performed at the same time as global search)
4. sort
5. chunk

Perhaps starting out with a separate component that uses the same configuration and just re-add one feature at a time and use observables for everything. 

I thought I'd try and draw a sketch of the flow using draw.io or something just to make the architecture clearer but that will have to wait a while I'm afraid as I'm going away on a holiday soon and won't be back until mid may. But if you have any suggestions feel free to share them, I'm might not be able to commit any code for a while but I should be able to check my mail from time to time and merge minor pull requests.

mklein994 commented:
OK thanks for the heads-up.

mklein994 committed:
Add rows shown filter and page indicator

mklein994 committed:
Include a row only if every column search string has a match

Empty column search strings are ignored. This means that the filter does not
apply to that column.

mklein994 committed:
Map rows through search/value function if defined

mklein994 committed:
Comment on custom search/value mapping for column search

mklein994 committed:
Use relative paths for column-search pipe imports

mklein994 committed:
Merge branch 'master' into column-search

mklein994 committed:
Merge branch 'master' into column-search

mklein994 committed:
Move column-search to new example structure

mklein994 committed:
Add a custom column search example component

danloveg committed:
Add table to Custom Column Search example

mklein994 committed:
Add an example search header component

mklein994 committed:
Add search header to examples for custom column search

mklein994 committed:
Use small caps for search header name

danloveg committed:
Search within column with custom header component

mklein994 committed:
Use a type definition for search functions

danloveg committed:
Add new event to custom component for column search

mklein994 committed:
Reindent custom header component in generic-table

mklein994 committed:
Use template reference for searchWithinColumn

Better than passing $event.

Also remove unused @ViewChild('myTable') reference.

mklein994 committed:
Remove unused inputBox @ViewChild element

mklein994 committed:
Give allRows a type in the column search pipe

mklein994 committed:
Refactor and document the column search pipe

Split into smaller functions to make it more manageable.

mklein994 committed:
Add pagination to custom column search example

mklein994 committed:
Remove type information from jsdocs in column search pipe

mklein994 committed:
Merge branch 'master' into column-search

Commit:
Filter out null values when searching a specific column

Fields with null values are included (for now) only when the search box
for that column is empty.

This fixes a bug where searching in a column that contains null values
would crash.

Commit:
Don't replace field with result of search function

Before, when a field defined a search function, the transformation
persisted beyond the column search pipe.

Now, the transformation is only done to confirm a match; the original
value for that field is returned.

Commit:
Add column search feature to show only null values

Commit:
Handle only null toggle in custom column search example

Commit:
Remove unused imports from column search examples

mklein994 committed:
Merge branch 'master' into column-search

Resolved conflict by adding a call to toggleAllRows() in the gt-checkbox
element.

Commit:
Merge branch 'master' into column-search

Resolved conflicts by moving the generic table template changes to the
new file.

Commit:
Merge branch 'master' into column-search

Commit:
Merge branch 'master' into column-search

Commit:
Merge branch 'master' into column-search

Commit:
Merge branch 'master' into column-search

hjalmers commented:
Closing: `angular-generic-table` has been rebuilt from the ground up for v5.0.0 (Angular 21, standalone components, signals). The default branch is now `main` with an entirely new history, so this PR no longer applies cleanly. Thanks for the contribution — please re-open against `main` if the change is still relevant on the new codebase.
